Add new awsconfig creds

jnpacker opened the pull request:
This allows the hypershift CLI infra AWS to work with key and secretKey credential values, when integrated into controllers or other environments where the credentials file may not be available.

This is to address issue: #859 

This adds AWSKey and AWSSecretKey to the options we use for infra.aws.create and infra.aws.create_iam.  The same was added to the destroy.  This allows someone consuming the Go functions to include the credentials, but leaves the secret definition, access to the impelmenter

jnpacker commented:
/cc @csrwng

csrwng commented on cmd/infra/aws/create.go:
```diff
@@ -19,6 +19,8 @@ type CreateInfraOptions struct {
 	InfraID            string
 	AWSCredentialsFile string
+	AWSKey             string
+	AWSSecretKey       string
```
should there be flags for these? Also if there are, AWSCreds should not be required (https://github.com/openshift/hypershift/pull/860/files#diff-718f3a535d5069f65c655a6143748ec6b3ac901d75f86e8ce59fe79fad4078ceR77). You may also want to add some validation that you can specify either/or but not both.

jnpacker replied:
That's the question I had. I don't know if we want to add two flags per provider for credentials.  All three main cloud providers have the local file.  The reason this was introducing this was if you want to call infra.aws.create or infra.aws.create_iam from external go code, in a controller, there is no way to have the creds file, but now you can dynamically inject the static cred's in this usecase.

If we feel we want to add flags for these though, I can add them in.

@csrwng does t

enxebre replied:
I think I'd rather avoid to enable users to pass text plain creds through the cli, specially since we don't have a need for it atm.

csrwng replied:
@enxebre agreed. I didn't initially understand that this code is being vendored and called directly. That being the case, I see no need to expose externally.

csrwng commented on cmd/infra/aws/destroy.go:
```diff
@@ -31,6 +31,8 @@ type DestroyInfraOptions struct {
 	InfraID            string
 	AWSCredentialsFile string
+	AWSKey             string
+	AWSSecretKey       string
```
same here, seems that you'd need to expose these via flags?

csrwng reviewed (commented):
Thanks @jnpacker some initial comments

enxebre commented on cmd/infra/aws/util/util.go:
```diff
@@ -31,12 +35,23 @@ func NewConfig(credentialsFile, region string) *aws.Config {
+func NewAWSConfig(credentialsFile string, credKey string, credSecretKey string, region string) *aws.Config {
+
+	if credentialsFile == "" {
+		return newConfig(credentials.NewStaticCredentials(credKey, credSecretKey, ""), region)
```
why do we need to have 2 funcs with the same name?
Can't we just have a single unit tested function i.e NewAWSConfig?

jnpacker replied:
I didn't want to replace NewConfig everywhere its used as I only need access to those four methods. 
I can expand its use to everywhere or change NewAWSConfig to something like `multiCredAWSConfig`?

enxebre replied:
ok, let's just update NewConfig to support the new params and have a single consistent way to create the aws config we use everywhere and behaves deterministically for any given input.
FWIW could also use provider chain https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/#ChainProvider

enxebre commented:
So this is to enable the use case where a higher level controller hardcodes the secret key in code in a `CreateInfraOptions` definition and pass it through `infra.aws.create` programatically? If so I don't think we should enable that as it seems to go against best practices. Why is a file creds and env vars for dev not a fair/sufficient assumption? by the ref issue I'm not really able to see the need for the change. Could you elaborate?

jnpacker commented:
We have a controller, that reads a Provider secret we have that has the Cloud Provider credential inside it, the controller borrows the CLI code, vs us just rewriting it, so the controller can provision the infrastructure pieces, like the CLI.  We can't mount the secret, as its a controller pod and not a run once job, and the secret reference changes.

This change allows us an entry point into the CLI code.

ACM flow:
1. A CR is created, it contains some HostedCluster and NodePool details (their specs) and whether we want to configure AWS (this pull allows us to directly leverage the CLI code)
2. If they choose to configure AWS infrastructure, the controller uses the secret they provided for the AWS credential and enters into the CLI code at the points in this Pull request. create.go and create_iam.go. 
3. When they delete the CR, first the NodePool then the HostedCluster  is cleaned up, and finally the destory.go and destroy_iam.go are called using the secret.

In ACM/MCE this allows credentials to be easily added, removed and updated across hypershift managementCluster ACM manages.

enxebre commented:
Thanks for the detailed context @jnpacker

jnpacker commented on cmd/infra/aws/util/util.go:
```diff
@@ -31,12 +35,23 @@ func NewConfig(credentialsFile, region string) *aws.Config {
+	return awsRoute53Config
+}
+
+func NewAWSConfig(credentialsFile string, credKey string, credSecretKey string, region string) *aws.Config {
```
I fixed the name a bit and added some comments to make it more clear what is going on.

jnpacker commented:
Took me a couple of tries to squash it.  Doesn't PROW squash when it merges?

enxebre commented:
Thanks!

>Took me a couple of tries to squash it. Doesn't PROW squash when it merges?

GH has an option to squash on merge, still the messages for your local git history would remain which have no value in the main git history. Also squashing the PR in one single commit is not always desirable. So by convention for PRs we require atomic commits via `git ci -m"what" -m"why"`. FWIW GH recognises the second -m"" and add it as the PR desc automatically.

@jnpacker Other than https://github.com/openshift/hypershift/pull/860/files#r788825660 lgtm. I suggest we just update the existing NewConfig to support the new input and we use the NewConfig everywhere and keep code base simple with a single way to create aws config, we don't need more flexibility so no more funcs.

jnpacker commented:
@enxebre That works for me, I will move it to NewConfig and squash onc emore.

jnpacker commented:
@enxebre Done. Just waiting on build.

sjenning commented:
Build failed
```
test/e2e/autorepair_test.go:88:32: not enough arguments in...thub.com/openshift/hypershift/cmd/infra/aws/util".NewConfig
	have (string, string)
	want (string, string, string, string)
```

enxebre commented on cmd/infra/aws/util/util.go:
```diff
@@ -31,12 +32,26 @@ func NewConfig(credentialsFile, region string) *aws.Config {
+	return awsRoute53Config
+}
+
+// NewMultiCredsAWSConfig allows support for a CredentialsFile or StaticCredential (UserKey & UserSecretKey)
```
> // NewMultiCredsAWSConfig...

Comment needs to be updated

enxebre commented on cmd/infra/aws/util/util.go:
```diff
@@ -31,12 +32,26 @@ func NewConfig(credentialsFile, region string) *aws.Config {
+// NewConfig allows support for a CredentialsFile or StaticCredential (UserKey & UserSecretKey)
+// [Default] if all values are provided is to use credentialsFile
+// This allows methods to be used by the CLI or vendored for other Go code
+func NewConfig(credentialsFile string, credKey string, credSecretKey string, region string) *aws.Config {
```
fwiw I was thinking of a single function

```
func NewConfig(credentialsFile string, credKey string, credSecretKey string, region string) *aws.Config {
	var creds *credentials.Credentials
	creds = credentials.NewStaticCredentials(credKey, credSecretKey, "")
	if credentialsFile != "" {
		creds = credentials.NewSharedCredentials(credentialsFile, "default")
	}

	awsConfig := aws.NewConfig().WithRegion(region).WithCredentials(creds)
	awsConfig.Retryer = client.DefaultRetryer{
		NumMaxRetries:    10,
		MinRetryDelay:    5 * time.Second,
		MinThrottleDelay: 5 * time.Second,
	}

	return awsConfig
}
```

but yours impl works just fine as well

jnpacker replied:
I'll combing both this and the suggestion above.

csrwng commented:
@jnpacker looks like you need to rebase. Once you do, I can tag it

csrwng commented:
/lgtm

jnpacker commented:
/test all

jnpacker commented:
/retest

jnpacker commented:
/retest

I guess I keep retrying till the clusterClaim in e2e works?
